File: src/main/java/dev/klomptech/jbark/parser/AstNode.java
package dev.klomptech.jbark.parser;

import dev.klomptech.jbark.print.PrintStyle;
import java.util.List;

// One statement in a .woof program; line() is the source line for error messages only
public sealed interface AstNode
    permits AstNode.Program,
        AstNode.Print,
        AstNode.Listen,
        AstNode.Exit,
        AstNode.Wait,
        AstNode.Assign,
        AstNode.SetAttribute,
        AstNode.IfChain,
        AstNode.WhileLoop,
        AstNode.UntilLoop,
        AstNode.ForEach,
        AstNode.FunctionDef,
        AstNode.Break,
        AstNode.Continue,
        AstNode.StashInit,
        AstNode.StashAppend,
        AstNode.StashSet,
        AstNode.StashRemove,
        AstNode.StashClear,
        AstNode.PileInit,
        AstNode.PilePush,
        AstNode.PilePop,
        AstNode.PileClear,
        AstNode.Share,
        AstNode.Pass,
        AstNode.SetTrait,
        AstNode.TakeFromStash,
        AstNode.LineGroup {

  int line();

  record Program(List<AstNode> statements) implements AstNode {
    @Override
    public int line() {
      return 1;
    }
  }

  record Print(
      PrintStyle style, List<ParseExpression> values, int line, boolean pawPrint, String dogSpeaker)
      implements AstNode {
    public Print(final PrintStyle style, final List<ParseExpression> values, final int line) {
      this(style, values, line, false, null);
    }

    public Print(
        final PrintStyle style,
        final List<ParseExpression> values,
        final int line,
        final boolean pawPrint) {
      this(style, values, line, pawPrint, null);
    }
  }

  record Listen(String dog, int line) implements AstNode {}

  record Exit(int line) implements AstNode {}

  record Wait(double seconds, int line) implements AstNode {}

  record Assign(String variable, ParseExpression value, int line) implements AstNode {}

  record SetAttribute(String subject, String topic, ParseExpression value, int line)
      implements AstNode {}

  record IfBranch(ParseExpression condition, List<AstNode> steps) {}

  record IfChain(List<IfBranch> branches, List<AstNode> elseSteps, int line) implements AstNode {}

  record WhileLoop(ParseExpression condition, List<AstNode> steps, int line) implements AstNode {}

  record UntilLoop(ParseExpression condition, List<AstNode> steps, int line) implements AstNode {}

  record ForEach(String variable, String stash, List<AstNode> steps, int line) implements AstNode {}

  record FunctionDef(
      String name,
      List<String> params,
      List<AstNode> steps,
      ParseExpression returnExpression,
      int line)
      implements AstNode {}

  record Break(int line) implements AstNode {}

  record Continue(int line) implements AstNode {}

  record StashInit(String stash, List<ParseExpression> items, int line) implements AstNode {}

  record StashAppend(String stash, ParseExpression item, int line) implements AstNode {}

  record StashSet(String stash, String which, ParseExpression value, int line) implements AstNode {}

  record StashRemove(String stash, String which, int line) implements AstNode {}

  record StashClear(String stash, int line) implements AstNode {}

  record PileInit(String pile, List<ParseExpression> items, int line) implements AstNode {}

  record PilePush(String pile, ParseExpression item, int line) implements AstNode {}

  record PilePop(String pile, int line) implements AstNode {}

  record PileClear(String pile, int line) implements AstNode {}

  record Share(String giver, String stash, List<String> recipients, int line) implements AstNode {}

  record Pass(String giver, String recipient, int line) implements AstNode {}

  record SetTrait(String subject, String trait, boolean enabled, int line) implements AstNode {}

  record TakeFromStash(String subject, String stash, String which, int line) implements AstNode {}

  // Multiple statements parsed from one story line (example: whimper + take)
  record LineGroup(List<AstNode> statements, int line) implements AstNode {}
}

File: pbark/pbark/parser/ast_node.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from pbark.parser.parse_expression import ParseExpression
from pbark.print_style import PrintStyle


@dataclass
class Program:
    statements: list[Any]
    line: int = 1


@dataclass
class Print:
    style: PrintStyle
    values: list[ParseExpression]
    line: int
    paw_print: bool = False
    dog_speaker: str | None = None


@dataclass
class Listen:
    dog: str
    line: int


@dataclass
class Exit:
    line: int


@dataclass
class Wait:
    seconds: float
    line: int


@dataclass
class Assign:
    variable: str
    value: ParseExpression
    line: int


@dataclass
class SetAttribute:
    subject: str
    topic: str
    value: ParseExpression
    line: int


@dataclass
class IfBranch:
    condition: ParseExpression
    steps: list[Any]


@dataclass
class IfChain:
    branches: list[IfBranch]
    else_steps: list[Any] | None
    line: int


@dataclass
class WhileLoop:
    condition: ParseExpression
    steps: list[Any]
    line: int


@dataclass
class UntilLoop:
    condition: ParseExpression
    steps: list[Any]
    line: int


@dataclass
class ForEach:
    variable: str
    stash: str
    steps: list[Any]
    line: int


@dataclass
class FunctionDef:
    name: str
    params: list[str]
    steps: list[Any]
    return_expression: ParseExpression
    line: int


@dataclass
class Break:
    line: int


@dataclass
class Continue:
    line: int


@dataclass
class StashInit:
    stash: str
    items: list[ParseExpression]
    line: int


@dataclass
class StashAppend:
    stash: str
    item: ParseExpression
    line: int


@dataclass
class StashSet:
    stash: str
    which: str
    value: ParseExpression
    line: int


@dataclass
class StashRemove:
    stash: str
    which: str
    line: int


@dataclass
class StashClear:
    stash: str
    line: int


@dataclass
class PileInit:
    pile: str
    items: list[ParseExpression]
    line: int


@dataclass
class PilePush:
    pile: str
    item: ParseExpression
    line: int


@dataclass
class PilePop:
    pile: str
    line: int


@dataclass
class PileClear:
    pile: str
    line: int


@dataclass
class Share:
    giver: str
    stash: str | None
    recipients: list[str]
    line: int


@dataclass
class Pass:
    giver: str
    recipient: str
    line: int


@dataclass
class SetTrait:
    subject: str
    trait: str
    enabled: bool
    line: int


@dataclass
class TakeFromStash:
    subject: str
    stash: str
    which: str
    line: int


@dataclass
class LineGroup:
    statements: list[Any]
    line: int


AstNode = Any
